src/Extents.jl

```diff
@@ -60,4 +80,89 @@ function extent end
 extent(extent) = nothing
 extent(extent::Extent) = extent
 
+"""
+    intersects(ext1::Extent, ext2::Extent)
+
+Check if two `Extent` objects intersect.
+
+Returns `true` if the extents of all common dimensions share some values
+including just the edge values of thier range.
+
+Dimensions that are not shared are ignored. The order of dimensions is also ignored.
+
+If there are no common dimensions, `false` is returned.
+"""
+function intersects(ext1::Extent, ext2::Extent)
+    keys = _shared_keys(ext1, ext2)
+    if length(keys) == 0
+        return false # Otherwise `all` returns `true` for empty tuples
+    else
+        dimintersections = map(keys) do k
+            _bounds_intersect(ext1[unwrap(k)], ext2[unwrap(k)])
+        end
+        return all(dimintersections)
+    end
+end
+intersects(obj1, obj2) = intersects(extent(obj1), extent(obj2))
+intersects(obj1::Extent, obj2::Nothing) = false
+intersects(obj1::Nothing, obj2::Extent) = false
+intersects(obj1::Nothing, obj2::Nothing) = false
+
+"""
+    union(ext1::Extent, ext2::Extent)
+
+Get the union of two extents, e.g. the combined extent of both objects
+for all dimensions.
+"""
+function union(ext1::Extent, ext2::Extent)
+    keys = _shared_keys(ext1, ext2)
+    values = map(keys) do k
+        k = unwrap(k)
+        k_exts = (ext1[k], ext2[k])
+        a = min(map(first, k_exts)...)
+        b = max(map(last, k_exts)...)
+        (a, b)
+    end
+    return Extent{map(unwrap, keys)}(values)
+end
+union(obj1, obj2) = union(extent(obj1), extent(obj2))
+union(obj1, obj2, obj3, objs...) = union(union(obj1, obj2), obj3, objs...)
+
+"""
+    intersect(ext1::Extent, ext2::Extent)
+
+Get the intersect of two extents as another `Extent`, e.g.
+the area covered by the shared dimensions for both extents.
+
+If there is no intersection for any shared dimension, `nothing` will be returned.
+"""
+function intersect(ext1::Extent, ext2::Extent)
+    intersects(ext1, ext2) || return nothing
+    keys = _shared_keys(ext1, ext2)
+    values = map(keys) do k
+        k = unwrap(k)
+        k_exts = (ext1[k], ext2[k])
+        a = max(map(first, k_exts)...)
+        b = min(map(last, k_exts)...)
+        (a, b)
+    end
+    return Extent{map(unwrap, keys)}(values)
+end
+intersect(obj1, obj2) = intersect(extent(obj1), extent(obj2))
+intersect(obj1, obj2, obj3, objs...) = intersect(intersect(obj1, obj2), obj3, objs...)
+
+function _bounds_intersect(b1::Tuple, b2::Tuple)
+    (b1[1] <= b2[2] && b1[2] >= b2[1])
+end
+
+function _shared_keys(ext1::Extent{K1}, ext2::Extent{K2}) where {K1,K2}
+    reduce(K1; init=()) do acc, k
+        # Use a static Val{k} here instead of a Symbol
```

Comment:
This comment is solved, so can be removed. I'd appreciate it if we explain the black magic that's going on here in a comment.

Reply:
Ok I'll rewrite to explain a little more. Basically its forcing the symbol keys (that we know as constants) to stay in the type domain, so the namedtuple lookups are resolved at compile time. Otherwise we lose constant propagation in `map` so the symbols have to be looked up at runtime. That takes 10x longer than the actual extent comparisons.